pkg/rslog/logger_impl.go

```diff
@@ -140,34 +144,87 @@ func getLevel(level LogLevel) logrus.Level {
 	}
 }
 
+func (l *LoggerImpl) Buffer() {
+	l.CoreLoggerImpl = NewBufLogger(l.CoreLogger, l.Flush)
+	if l.wrappers == nil {
+		l.wrappers = make(map[*logrusEntryWrapper]struct{})
+	}
+}
+
+func (l *LoggerImpl) Flush() {
+	if bufLogger, ok := l.CoreLoggerImpl.(*BufLogger); ok {
+		storage := bufLogger.Storage
+
+		// Swap logger implementations and allow wrappers set to be garbage-collected.
+		l.CoreLoggerImpl = l.CoreLogger
+		for wrapper := range l.wrappers {
+			wrapper.CoreLoggerImpl = wrapper.coreLogger
+			// Allow this to be garbage-collected.
+			wrapper.wrappers = nil
+		}
+		// Allow this to be garbage-collected.
+		l.wrappers = nil
+
+		// Time to output retained logs to their respective loggers.
+		for _, entry := range storage.Logs {
+			lgr := entry.Logger
+
+			switch entry.Level {
+			case TraceLevel:
+				lgr.Tracef(entry.Message, entry.Args...)
+			case DebugLevel:
+				lgr.Debugf(entry.Message, entry.Args...)
+			case InfoLevel:
+				lgr.Infof(entry.Message, entry.Args...)
+			case WarningLevel:
+				lgr.Warnf(entry.Message, entry.Args...)
+			case ErrorLevel:
+				lgr.Errorf(entry.Message, entry.Args...)
+			}
+		}
+	}
+}
+
 func (l LoggerImpl) WithField(key string, value interface{}) Logger {
-	e := l.Logger.WithField(key, value)
-	return logrusEntryWrapper{Entry: e}
+	e := l.CoreLogger.WithField(key, value)
+	wrapper := &logrusEntryWrapper{coreLogger: e, CoreLoggerImpl: e, wrappers: l.wrappers}
+	if bufLogger, ok := l.CoreLoggerImpl.(*BufLogger); ok {
+		l.wrappers[wrapper] = struct{}{}
+		wrapper.CoreLoggerImpl = bufLogger.child(wrapper.coreLogger)
+	}
```

Comment:
WDYT about having a `wrapBufLoggerChild()` method to consolidate this check blocks across the `WithField(s)` methods?